packages/augur-core/src/contracts/para/AMMFactory.sol

```diff
@@ -0,0 +1,34 @@
+pragma solidity 0.5.15;
+
+import 'ROOT/libraries/CloneFactory2.sol';
+import 'ROOT/para/interfaces/IParaShareToken.sol';
+import 'ROOT/para/AMMExchange.sol';
+import 'ROOT/reporting/IMarket.sol';
+
+contract AMMFactory {
+    // market -> para -> amm
+    AMMExchange internal proxyToClone;
+
+    constructor(address _proxyToClone) public {
+        proxyToClone = AMMExchange(_proxyToClone);
+    }
+
+    function addAMM(IMarket _market, IParaShareToken _para, uint256 _fee) external returns (AMMExchange) {
+        address _amm = AMMExchange(createClone2(address(proxyToClone), salt(_market, _para)));
+        _amm.initialize(_market, _para, _fee);
```

Comment:
I recommend running the numbers on deployment cost vs usage costs.  By deploying like this, you can't have any immutable constants which means *every* call is going to be more expensive, and with the upcoming changes that increase storage reads to something like 2500 gas this is very likely to add up to a big number.

The AMM contract is small enough, that the increased gas cost in a direct deploy may be worth it since runtime costs will be substantially lower.